improver/calibration/rainforest_calibration.py

```diff
@@ -0,0 +1,170 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# (C) British Crown Copyright 2017-2021 Met Office.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""RainForests calibration Plugins."""
+
+import warnings
+
+import numpy as np
+from iris.cube import Cube, CubeList
+
+from improver import BasePlugin, PostProcessingPlugin
+
+# Passed to choose_set_of_percentiles to set of evenly spaced percentiles
+DEFAULT_ERROR_PERCENTILES_COUNT = 19
+DEFAULT_OUTPUT_REALIZATIONS_COUNT = 100
+
+
+class TrainRainForestsTreeModels(BasePlugin):
+    """Class to train tree models for use in RainForests calibration."""
+
+
+class ApplyRainForestsCalibration(PostProcessingPlugin):
+    """Class to calibrate input forecast given a series of RainForests tree models."""
+
+    def __init__(self, model_config_dict: dict, nthreads: int):
+        """Initialise the tree model variables used in the application of RainForests
+        Calibration.
+
+        Args:
+            model_config_dict:
+                Dictionary containing Rainforests model configuration variables.
+            nthreads:
+                Number of threads to use when initialising tree-model objects.
+
+        Dictionary is of format::
+
+            {
+                "-50.0" : {
+                    "lightgbm_model" : "<path_to_lightgbm_model_object>",
```

Comment:
Fair question. I should add that this is my first time using Mock/MonkeyPatch and it took some effort to get my head around what/how to test the __int__ function, so I completely understand the opacity.

The intention with the unit tests is to test that the relevant model objects are being initialised given the values passed via model_config; for the test we are not really interested in the result model tree initialisation, just that the relevant load function is being used. Consequently, the goal of mocking here is to override the lightgbm Booster and treelite Predictor __init__ functions to generate Mock instances of these objects. In doing this, we can test that the relevant load function is being called without actually having to load anything, which simplifies the tests as there is no need to create sample files to load.

So the tests should pass, despite the files that would be loaded not existing.

Reply:
Thanks for the explanation - I thought that might be what was happening, but you've made it a lot clearer for me!
What this method prevents, is testing that the true objects do initiate, and what happens when one of the file names is `None` (see separate comment). I think the best solution is to trap that potential `None` case so that you don't have to test the true behaviour of the external module.

Reply:
No worries. I tried the case where `lightgbm_model` is missing for one of the error-threshold keys and it fails at the point of initialising the associated `Booster` object as `None` is not a valid value. However, I agree that the best way to handle this is test for `None` before loading in values, so I've put in a check before initialising the `Booster` objects and raising a `ValueError` if any `None` values are in `lightgbm_model_filenames`.

I've also added a unit-test to check this case.